src/lib.rs

```diff
@@ -192,31 +195,84 @@ impl Uffd {
-        const MSG_SIZE: usize = mem::size_of::<raw::uffd_msg>();
-        let mut buf = [0x00u8; MSG_SIZE];
+        let mut buf = [0; UFFD_EVENT_SIZE];
+        let mut iter = self.read_events(&mut buf)?;
+        let event = iter.next().transpose()?;
+        assert!(iter.next().is_none());
+        Ok(event)
+    }
 
-        // read one uffd_msg at a time; maybe implement an iterator for handling many at once?
-        let res = read(self.as_raw_fd(), &mut buf);
+    /// Read multiple events from the userfaultfd object using the given buffer.
+    ///
+    /// The buffer size must be a non-zero multiple of `UFFD_EVENT_SIZE`.
+    ///
+    /// If the `Uffd` object was created with `non_blocking` set to `false`, this will block until
+    /// an event is successfully read or an error is returned.
+    ///
+    /// If `non_blocking` was `true`, this will immediately return an empty iterator if the file
+    /// descriptor is not ready for reading.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// fn read_events(uffd: &userfaultfd::Uffd) -> userfaultfd::Result<()> {
+    ///     // Read up to 100 events at a time
+    ///     let mut buf = [0; 100 * userfaultfd::UFFD_EVENT_SIZE];
+    ///     for event in uffd.read_events(&mut buf)? {
+    ///         let event = event?;
+    ///         // Do something with the event...
+    ///     }
+    ///     Ok(())
+    /// }
+    /// ```
+    pub fn read_events<'a>(
+        &self,
+        buf: &'a mut [u8],
+    ) -> Result<impl Iterator<Item = Result<Event>> + 'a> {
+        if buf.is_empty() || (buf.len() % UFFD_EVENT_SIZE) != 0 {
+            return Err(Error::SystemError(Errno::EINVAL.into()));
+        }
 
-        match res {
-            Err(e) if e.as_errno() == Some(Errno::EAGAIN) => return Ok(None),
-            Err(e) => Err(e)?,
+        let nread = match read(self.as_raw_fd(), buf) {
+            Err(e) if e.as_errno() == Some(Errno::EAGAIN) => 0,
+            Err(e) => return Err(Error::SystemError(e)),
             Ok(nread) => {
                 if nread == libc::EOF as usize {
                     return Err(Error::ReadEof);
                 }
-                if nread != MSG_SIZE {
+                let remainder = nread % UFFD_EVENT_SIZE;
+                if remainder != 0 {
                     return Err(Error::IncompleteMsg {
-                        read: nread,
-                        expected: MSG_SIZE,
+                        read: remainder,
+                        expected: UFFD_EVENT_SIZE,
                     });
                 }
-                let msg = unsafe { *(buf.as_mut_ptr() as *mut raw::uffd_msg) };
-                let event = Event::from_uffd_msg(&msg)?;
-                return Ok(Some(event));
+                nread
             }
-        }
+        };
+
+        Ok(buf[..nread].chunks(UFFD_EVENT_SIZE).map(|msg| {
+            let msg = unsafe { &*(msg.as_ptr() as *const raw::uffd_msg) };
```

Comment:
does this imply that the buffer chunks need to have the same alignment as a raw::uffd_msg? in C it would, but I'm not sure if that rule applies here.

Reply:
I'll double check this to ensure the size used here takes into account the alignment.

Reply:
I've pushed a new commit that changes `read_events` such that it doesn't take a slice of bytes.

This should make it less error prone for users too.

Let me know what you think about that change.